File: core/contractsdk/go/example/game_assets/game_assets.go
package main

import (
	"github.com/xuperchain/xuperchain/core/contractsdk/go/code"
	"github.com/xuperchain/xuperchain/core/contractsdk/go/driver"
	"github.com/xuperchain/xuperchain/core/contractsdk/go/utils"
	utils2 "github.com/xuperchain/xuperchain/core/contractsdk/go/utils"
)

type gameAssets struct {
}

const (
	ADMIN      = "admin"
	ASSETTYPE  = "AssetType_"
	USERASSET  = "UserAsset_"
	ASSET2USER = "Asset2User_"
)

func (ga *gameAssets) Initialize(ctx code.Context) code.Response {
	args := struct {
		Admin []byte `json:"admin",required:"true"`
	}{}

	if err := utils.Validate(ctx.Args(), &args); err != nil {
		return code.Error(err)
	}
	if err := ctx.PutObject(utils2.Concat(ADMIN), args.Admin); err != nil {
		return code.Error(err)
	} else {
		return code.OK(nil)
	}

}

//TODO @fengjin
func (ga *gameAssets) isAdmin(ctx code.Context, caller string) bool {
	admin, err := ctx.GetObject(utils2.Concat(ADMIN))
	if err != nil {
		//TODO 错误处理
		//return
	}
	return utils2.Compare(admin, caller) == 0
}

func (ga *gameAssets) AddAssetType(ctx code.Context) code.Response {
	caller := ctx.Initiator()
	if caller == "" {
		return code.Error(utils2.ErrMissingCaller)
	}
	if !ga.isAdmin(ctx, caller) {
		return code.Error(utils2.ErrPermissionDenied)
	}
	args := struct {
		TypeID   []byte `json:"typeid",required:"true"`
		TypeDesc []byte `json:"typedesc",required:"true"`
	}{}
	if err := utils.Validate(ctx.Args(), &args); err != nil {
		return code.Error(err)
	}
	assetKey := utils2.Concat(ASSETTYPE, args.TypeID)
	if err := utils2.CheckExist(ctx, assetKey); err != nil {
		return code.Error(err)
	}
	if err := ctx.PutObject(assetKey, args.TypeDesc); err != nil {
		return code.Error(err)
	}
	return code.OK(nil)
}

func (ga *gameAssets) ListAssetType(ctx code.Context) code.Response {
	start := utils2.Concat(ASSETTYPE)
	end := utils2.Concat(ASSETTYPE, "~")
	iter := ctx.NewIterator(start, end)
	result := []byte{}
	for iter.Next() {
		if len(iter.Key()) > len(ASSETTYPE) {
			//result = append(result, utils2.Concat(iter.Key()[:len([]byte(ASSETTYPE):],":",iter.Value(),"\n")))
		}
	}
	return code.OK(result)
}

func (ga *gameAssets) getAssetByUser(ctx code.Context) code.Response {
	//这里也不想搞了
	caller := ctx.Initiator()
	if caller == "" {
		return code.Error(utils2.ErrMissingCaller)
	}
	if !ga.isAdmin(ctx, caller) {
		return code.Error(utils2.ErrPermissionDenied)
	}
	args := struct {
		UserID []byte `json:"userid",required:"false"`
	}{}
	if err := utils.Validate(ctx.Args(), &args); err != nil {
		return code.Error(err)
	}
	//	TODO 这一段太重复了，想办法抽象一下吧

	//TODO @fengjin	空白slice 和nil以及slice的零值
	userId := []byte(caller)
	if args.UserID != nil && len(args.UserID) > 0 {
		userId = args.UserID
	}
	userAssetKey := utils2.Concat(USERASSET, userId, "_")
	start := userAssetKey
	end := utils2.Concat(start, "~")
	iter := ctx.NewIterator(start, end)
	result := []byte{}
	//迭代获取也好重复
	for iter.Next() {
		if len(iter.Key()) > len(userAssetKey) {
			assetId := iter.Key()[len(userAssetKey)]
			typeId := iter.Value()
			assetTypeKey := utils2.Concat(ASSETTYPE, typeId)
			if assetDesc, err := ctx.GetObject(assetTypeKey); err != nil {
				continue
				//	TODO @fengjin
			} else {
				result = append(result, utils2.Concat("assetid=", assetId, ",typeid=", typeId, ",assetDesc=", assetDesc, "\n")...)
			}
		}
	}
	return code.OK(result)
}

func (ga *gameAssets) NewAssetToUser(ctx code.Context) code.Response {
	caller := ctx.Initiator()
	if caller == "" {
		return code.Error(utils2.ErrMissingCaller)
	}
	if !ga.isAdmin(ctx, caller) {
		return code.Error(utils2.ErrPermissionDenied)
	}
	args := struct {
		UserId  []byte `json:"userid",required:"true"`
		TypeId  []byte `json:"typeid",required:"true"`
		AssetId []byte `json:"assetid",required:"true"`
	}{}
	if err := utils.Validate(ctx.Args(), &args); err != nil {
		return code.Error(err)
	}
	assetKey := utils2.Concat(ASSET2USER, args.AssetId)
	_, err := ctx.GetObject(assetKey)
	if err != nil {
		return code.Error(err)
	}
	userAssetKey := utils2.Concat(USERASSET, args.UserId, "_", args.AssetId)

	if err := ctx.PutObject(userAssetKey, args.TypeId); err != nil {
		return code.Error(err)
	}
	if err := ctx.PutObject(assetKey, args.UserId); err != nil {
		return code.Error(err)
	}
	return code.OK(args.AssetId)
}

func (ga *gameAssets) TradeAsset(ctx code.Context) code.Response {
	from := ctx.Initiator()
	if from == "" {
		return code.Error(utils2.ErrMissingCaller)
	}
	args := struct {
		To      []byte `json:"to",required:"true"`
		AssetId []byte `json:"assetid",required:"true"`
	}{}
	if err := utils.Validate(ctx.Args(), &args); err != nil {
		return code.Error(err)
	}
	userAssetKey := utils2.Concat(USERASSET, from, "_", args.AssetId)
	assetType, err := ctx.GetObject(userAssetKey)
	if err != nil {
		return code.Error(err)
	}
	if err := ctx.DeleteObject(userAssetKey); err != nil {
		return code.Error(err)
	}

	assetKey := utils2.Concat(ASSET2USER, args.AssetId)
	newuserAssetKey := utils2.Concat(USERASSET, args.To, "_", args.AssetId)
	if err := ctx.PutObject(newuserAssetKey, assetType); err != nil {
		return code.Error(err)
	}
	if err := ctx.PutObject(assetKey, args.To); err != nil {
		return code.Error(err)
	}
	return code.OK(args.AssetId)
}

func main() {
	driver.Serve(new(gameAssets))
}

File: core/contractsdk/java/example/game_assets/src/main/java/com/baidu/xuper/example/GameAssets.java
package com.baidu.xuper.example;

import java.math.BigInteger;
import java.util.HashMap;
import java.util.Map;

import com.baidu.xuper.Context;
import com.baidu.xuper.Contract;
import com.baidu.xuper.ContractMethod;
import com.baidu.xuper.Driver;
import com.baidu.xuper.Response;
import com.baidu.xuper.utils.ByteUtils;
import com.sun.tools.javac.util.ByteBuffer;

/**
 * Counter
 */
public class GameAssets implements Contract {
    final String ASSETTYPE = "AssetType_";
    final String USERASSET = "UserAsset";
    final String ASSET2USER = "Asset2User_";


    @Override
    @ContractMethod
    public Response initialize(Context ctx) {
        byte[] admin = ctx.args().get("admin");
        if (admin.length == 0) {
            return Response.error("missing admin address");
        }
        ctx.putObject("admin".getBytes(), admin);
        return Response.ok("ok".getBytes());
    }

    private boolean isAdmin(Context ctx) {
        return this.isAdmin(ctx, ctx.caller());
    }

    private boolean isAdmin(Context ctx, String caller) {
        byte[] admin = ctx.getObject("admin".getBytes());
        return ByteUtils.equal(admin, caller);
    }

    @ContractMethod
    public Response AddAssetType(Context ctx) {
        if (!this.isAdmin(ctx)) {
            return Response.error("only admin can add new asset type");
        }
        byte[] typeId = ctx.args().get("typeid");
        byte[] typeDesc = ctx.args().get("typedesc");// TODO @fengjin 向后兼容？？
        if (typeId.length == 0 || typeDesc.length == 0) {
            return Response.error("missing typeid or typedesc");
        }
        byte[] assetTypeKey = ByteUtils.concat(ASSETTYPE, typeId);
        if (ctx.getObject(assetTypeKey).length != 0) {
            return Response.error("type id does already exist");
        }
        ctx.putObject(assetTypeKey, typeDesc);
        return Response.ok(typeId);
    }

    @ContractMethod
    public Response listAssetType(Context ctx) {
        ByteBuffer buf = new ByteBuffer();
        ctx.newIterator(ASSETTYPE.getBytes(), ByteUtils.concat(ASSETTYPE, "~")).forEachRemaining(
                elem -> {
                    String assetType = ByteUtils.toString(elem.getKey());
                    buf.appendBytes((assetType.substring(ASSETTYPE.length()) + ":" +
                            elem.getValue() +
                            "\n").getBytes());// TODO @fengjin silce 的tostring 方法，类型转换规则
                }
        );
        return Response.ok(buf.elems);
    }

    @ContractMethod
    public Response getAssetsByUser(Context ctx) {
        //        admin cat get asset of any user
        String userId = ctx.caller();
        if (isAdmin(ctx, userId)) {
            if (ctx.args().get("userid").length != 0) { // TODO 参数格式
                userId = ByteUtils.toString(ctx.args().get("user_id"));
            }
        }
        byte[] userAsstKey = ByteUtils.concat(ByteUtils.concat(USERASSET, userId), "_");
        byte[] start = userAsstKey;
        byte[] end = ByteUtils.concat(start, "~");
//         这里再封装一层？？传递一个函数？？
        ByteBuffer buf = new ByteBuffer();
        ctx.newIterator(start, end).forEachRemaining(
                elem -> {
                    if (elem.getKey().length > userAsstKey.toString().length()) { // TODO why
                        String assetId = ByteUtils.toString(elem.getKey()).substring(userAsstKey.length);
                        String typeId = ByteUtils.toString(elem.getValue());
                        String assetTypeKey = ASSETTYPE + typeId;
                        String assetDesc = ByteUtils.toString(ctx.getObject(assetTypeKey.getBytes()));    // TODO 检查？？
                        buf.appendBytes(("assetid=" + assetId + ",typeid=" + typeId + ",assetDesc=" + assetDesc + "\n").getBytes());
                    }
                }
        );
        return Response.ok(buf.elems);
    }

    @ContractMethod
    public Response newAssetToUser(Context ctx) {
        if (!this.isAdmin(ctx)) {
            return Response.error("only admin can add new asset type");
        }
        byte[] userId = ctx.args().get("user_id");
        byte[] typeId = ctx.args().get("type_id");
        byte[] assetId = ctx.args().get("asset_id");
        if (userId.length == 0 || typeId.length == 0 || assetId.length == 0) {
            return Response.error("missing user_id or type_id or asst_id");
        }
        byte[] assetKey = ByteUtils.concat(ASSET2USER, assetId);
//        这里的检查是需要的么？？
        if (ctx.getObject(assetKey).length > 0) {
            return Response.error("asset id already exist");
        }

        byte[] userAssetKey = (USERASSET + ByteUtils.toString(userId) + "_" + ByteUtils.toString(assetId)).getBytes();
        ctx.putObject(userAssetKey, typeId);
        ctx.putObject(assetKey, userId);
        return Response.ok(assetId);
    }

    public Response tradeAsset(Context ctx) {
        byte[] from = ctx.caller().getBytes();// TODO ok?
        if (from.length == 0) {
            return Response.error("missing initiator"); // TODO why?
        }
        byte[] to = ctx.args().get("to");
        byte[] assetId = ctx.args().get("assetid"); // TODO??
        if (to.length == 0 || assetId.length == 0) {
            return Response.error("missing to or assetid");
        }
        byte[] userAssetKey = (USERASSET + ByteUtils.toString(from) + "_" + ByteUtils.toString(assetId)).getBytes();
        byte[] assetType = ctx.getObject(userAssetKey);
        if (assetType.length == 0) {
            return Response.error("you do not have asset with id " + ByteUtils.toString(assetId));
        }
        ctx.deleteObject(userAssetKey);
        byte[] assetKey = ByteUtils.concat(ASSET2USER, assetId);
        byte[] newUserAssetKey = (USERASSET + ByteUtils.toString(to) + "_" + ByteUtils.toString(assetId)).getBytes();
        ctx.putObject(newUserAssetKey, assetId);
        ctx.putObject(assetKey, to);
        return Response.ok(assetId);
    }

    public static void main(String[] args) {
        Driver.serve(new GameAssets());
    }
}